packages/dashboard-message-bus/lib/DashboardMessageBus.ts

```diff
@@ -0,0 +1,194 @@
+import WebSocket from "isomorphic-ws";
+import delay from "delay";
+import { EventEmitter } from "events";
+import {
+  base64ToJson,
+  broadcastAndAwaitFirst,
+  broadcastAndDisregard,
+  createMessage,
+  jsonToBase64,
+  startWebSocketServer
+} from "./utils";
+import { promisify } from "util";
+
+interface UnfulfilledRequest {
+  publisher: WebSocket;
+  data: WebSocket.Data;
+}
+
+export class DashboardMessageBus extends EventEmitter {
+  private publishServer: WebSocket.Server;
+  private subscribeServer: WebSocket.Server;
+  private publishers: WebSocket[] = [];
+  private subscribers: WebSocket[] = [];
+
+  private unfulfilledRequests: Map<string, UnfulfilledRequest> = new Map([]);
+
+  constructor(
+    public publishPort: number,
+    public subscribePort: number,
+    public host: string = "localhost"
+  ) {
+    super();
+  }
+
+  async start() {
+    this.subscribeServer = await startWebSocketServer({
+      host: this.host,
+      port: this.subscribePort
+    });
+
+    this.subscribeServer.on("connection", (newSubscriber: WebSocket) => {
+      newSubscriber.on("close", () => {
+        this.logToPublishers("Subscriber disconnected", "connections");
+
+        this.subscribers = this.subscribers.filter(
+          subscriber => subscriber !== newSubscriber
+        );
+
+        this.terminateIfNoConnections();
+      });
```

Comment:
As my proposed alternative I refactored the add/remove logic into separate methods so that it's clearer what is happening in this callback without having to add extra comments. I *did* add a jsdoc comment to these new functions with some notes.